File: community_discussions/trustpilot_data/scrape.py
import requests
import json
import csv
import datetime
import time
from bs4 import BeautifulSoup

def scrape_trustpilot_reviews(domain="nykaafashion.com", pages=5):
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.5",
    }
    
    all_reviews = []
    
    print(f"Scraping Trustpilot reviews for {domain}...")
    
    for page in range(1, pages + 1):
        url = f"https://www.trustpilot.com/review/{domain}?page={page}"
        print(f"Fetching page {page}: {url}")
        
        try:
            response = requests.get(url, headers=headers)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, 'html.parser')
            
            # Trustpilot embeds review data in application/ld+json scripts for SEO
            scripts = soup.find_all('script', type='application/ld+json')
            page_reviews = []
            
            for script in scripts:
                if not script.string:
                    continue
                try:
                    data = json.loads(script.string)
                    # Data can be a list or a dict
                    if isinstance(data, list):
                        for item in data:
                            if item.get('@type') == 'LocalBusiness' and 'review' in item:
                                for review in item['review']:
                                    page_reviews.append(parse_ld_json_review(review))
                    elif isinstance(data, dict):
                        if data.get('@type') == 'LocalBusiness' and 'review' in data:
                            for review in data['review']:
                                page_reviews.append(parse_ld_json_review(review))
                except json.JSONDecodeError:
                    continue
            
            if page_reviews:
                all_reviews.extend(page_reviews)
                print(f"Found {len(page_reviews)} reviews on page {page}.")
            else:
                print(f"No reviews found on page {page}. (May have reached the end or blocked by anti-bot)")
                break
                
            time.sleep(1.5) # Be polite to the server
            
        except requests.exceptions.RequestException as e:
            print(f"Error fetching page {page}: {e}")
            break
            
    return all_reviews

def parse_ld_json_review(review):
    return {
        'platform': 'Trustpilot',
        'author': review.get('author', {}).get('name', 'Unknown'),
        'rating': review.get('reviewRating', {}).get('ratingValue'),
        'headline': review.get('headline', ''),
        'content': review.get('reviewBody', ''),
        'date_published': review.get('datePublished', ''),
        'scraped_at': datetime.datetime.now().isoformat()
    }

if __name__ == "__main__":
    # We will scrape the first 5 pages as an example (20 reviews per page -> ~100 reviews)
    reviews = scrape_trustpilot_reviews("nykaafashion.com", pages=5)
    
    if reviews:
        print(f"\nSuccessfully scraped {len(reviews)} reviews.")
        
        # Save to JSON
        json_filename = 'nykaa_trustpilot_reviews.json'
        with open(json_filename, 'w', encoding='utf-8') as f:
            json.dump(reviews, f, indent=4, ensure_ascii=False)
        print(f"Data saved to {json_filename}")
        
        # Save to CSV
        csv_filename = 'nykaa_trustpilot_reviews.csv'
        keys = reviews[0].keys()
        with open(csv_filename, 'w', newline='', encoding='utf-8') as f:
            dict_writer = csv.DictWriter(f, fieldnames=keys)
            dict_writer.writeheader()
            dict_writer.writerows(reviews)
        print(f"Data saved to {csv_filename}")
    else:
        print("\nNo reviews found or scraping failed.")

File: community_discussions/trustpilot_data/scrape.js
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());
const fs = require('fs');
const path = require('path');

async function scrapeTrustpilot() {
    const domain = 'nykaafashion.com';
    const pages = 5;
    const allReviews = [];

    console.log(`Scraping Trustpilot reviews for ${domain}...`);

    const browser = await puppeteer.launch({ 
        headless: 'new',
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--window-size=1920,1080']
    });
    
    const page = await browser.newPage();
    await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
    await page.setViewport({ width: 1920, height: 1080 });
    
    for (let p = 1; p <= pages; p++) {
        const url = `https://www.trustpilot.com/review/${domain}?page=${p}`;
        console.log(`Fetching page ${p}: ${url}`);
        
        try {
            await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 60000 });
            
            // Trustpilot uses Cloudflare / Datadome, sometimes it asks for a captcha
            const content = await page.content();
            if (content.includes('cf-browser-verification') || content.includes('captcha')) {
                console.warn(`WARNING: Cloudflare/Captcha challenge detected on page ${p}.`);
            }
            
            // Extract reviews from the page DOM and LD+JSON
            const reviews = await page.evaluate(() => {
                const pageReviews = [];
                
                // Method 1: Try structured data first
                const scripts = document.querySelectorAll('script[type="application/ld+json"]');
                scripts.forEach(script => {
                    if (!script.innerText) return;
                    try {
                        const data = JSON.parse(script.innerText);
                        if (Array.isArray(data)) {
                            data.forEach(item => {
                                if (item['@type'] === 'LocalBusiness' && item.review) {
                                    item.review.forEach(r => pageReviews.push(r));
                                }
                            });
                        } else if (data['@type'] === 'LocalBusiness' && data.review) {
                            data.review.forEach(r => pageReviews.push(r));
                        }
                    } catch (e) {}
                });
                
                let parsedReviews = pageReviews.map(review => ({
                    platform: 'Trustpilot',
                    author: review.author ? review.author.name : 'Unknown',
                    rating: review.reviewRating ? review.reviewRating.ratingValue : null,
                    headline: review.headline || '',
                    content: review.reviewBody || '',
                    date_published: review.datePublished || '',
                    scraped_at: new Date().toISOString()
                }));

                // Method 2: If structured data failed, fallback to DOM parsing
                if (parsedReviews.length === 0) {
                    const reviewCards = document.querySelectorAll('article, .paper_paper__1PY90');
                    reviewCards.forEach(card => {
                        const author = card.querySelector('.typography_heading-xxs__qkLXz, [data-consumer-name-typography]')?.innerText?.trim() || 'Unknown';
                        const headline = card.querySelector('h2, [data-service-review-title-typography]')?.innerText?.trim() || '';
                        const content = card.querySelector('p[data-service-review-text-typography]')?.innerText?.trim() || '';
                        const dateText = card.querySelector('time')?.getAttribute('datetime') || '';
                        
                        // Rating is usually in an img alt tag or a specific div
                        const ratingImg = card.querySelector('img[alt*="Rated"]');
                        let rating = null;
                        if (ratingImg) {
                            const match = ratingImg.alt.match(/Rated (\\d)/);
                            if (match) rating = match[1];
                        }
                        
                        if (content || headline) {
                            parsedReviews.push({
                                platform: 'Trustpilot',
                                author,
                                rating,
                                headline,
                                content,
                                date_published: dateText,
                                scraped_at: new Date().toISOString()
                            });
                        }
                    });
                }
                
                return parsedReviews;
            });
            
            if (reviews.length > 0) {
                allReviews.push(...reviews);
                console.log(`Found ${reviews.length} reviews on page ${p}.`);
            } else {
                console.log(`No reviews found on page ${p}. End of results or bot detection.`);
                break;
            }
            
            // Random delay to mimic human behavior
            await new Promise(resolve => setTimeout(resolve, 2000 + Math.random() * 1500));
            
        } catch (err) {
            console.error(`Error on page ${p}: ${err.message}`);
            break;
        }
    }
    
    await browser.close();
    
    if (allReviews.length > 0) {
        console.log(`\nSuccessfully scraped ${allReviews.length} reviews.`);
        
        const jsonFilename = path.join(__dirname, 'nykaa_trustpilot_reviews.json');
        fs.writeFileSync(jsonFilename, JSON.stringify(allReviews, null, 4));
        console.log(`Data saved to ${jsonFilename}`);
        
    } else {
        console.log('\nNo reviews found or scraping failed.');
    }
}

scrapeTrustpilot();
